substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfSectionImpl.java

```diff
@@ -300,105 +286,147 @@ protected int putUTF8StringBytes(String s, int startChar, byte[] buffer, int p)
      * Common write methods that check for a null buffer.
      */
 
-    protected void patchLength(int lengthPos, byte[] buffer, int pos) {
+    protected int writeByte(byte b, byte[] buffer, int p) {
         if (buffer != null) {
-            int length = pos - (lengthPos + 4);
-            putInt(length, buffer, lengthPos);
+            return putByte(b, buffer, p);
+        } else {
+            return p + 1;
         }
     }
 
-    protected int writeAbbrevCode(long code, byte[] buffer, int pos) {
-        if (buffer == null) {
-            return pos + putSLEB(code, scratch, 0);
+    protected int writeShort(short s, byte[] buffer, int p) {
+        if (buffer != null) {
+            return putShort(s, buffer, p);
         } else {
-            return putSLEB(code, buffer, pos);
+            return p + 2;
         }
     }
 
-    protected int writeTag(long code, byte[] buffer, int pos) {
-        if (buffer == null) {
-            return pos + putSLEB(code, scratch, 0);
+    protected int writeInt(int i, byte[] buffer, int p) {
+        if (buffer != null) {
+            return putInt(i, buffer, p);
         } else {
-            return putSLEB(code, buffer, pos);
+            return p + 4;
```

Comment:
Minor nit: wouldn't it be clearer to use those `BYTES` fields of `Short`, `Integer`, etc.?
```suggestion
            return p + Integer.BYTES;
```

Reply:
Yes, it would be a tad clearer. However, I don't think Integer.Bytes is ever going to be anything other than 4 and I suspect (but cannot guarantee) that every Java programmer knows that. So I'm not sure it matters too much.

Also if you look at the if branch for this else branch you will see that it calls `putInt` which is defined as

```
    protected int putInt(int i, byte[] buffer, int p) {
        int pos = p;
        if (littleEndian()) {
            buffer[pos++] = (byte) (i & 0xff);
            buffer[pos++] = (byte) ((i >> 8) & 0xff);
            buffer[pos++] = (byte) ((i >> 16) & 0xff);
            buffer[pos++] = (byte) ((i >> 24) & 0xff);
        } else {
            buffer[pos++] = (byte) ((i >> 24) & 0xff);
            buffer[pos++] = (byte) ((i >> 16) & 0xff);
            buffer[pos++] = (byte) ((i >> 8) & 0xff);
            buffer[pos++] = (byte) (i & 0xff);
        }
        return pos;
    }
```

So, the value 4 is *kind of hard-wired* into the code.

I suppose I could change `putInt` to iterate `for (i = 0; i < Integer.BYTES; i++)` and use `8 * i` or `32 - (8 * i)` as the shift operand and so on but ...  well, life is short ;-)

Reply:
The PR already got integrated. Feel free to add some asserts as part of one of your next PRs :smiley:.